File: avro/src/main/java/com/intenthq/pucket/javacompat/avro/AvroPucketInstantiator.java
package com.intenthq.pucket.javacompat.avro;

import org.apache.avro.generic.IndexedRecord;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;

import com.intenthq.pucket.Pucket;
import com.intenthq.pucket.avro.AvroPucket;
import com.intenthq.pucket.avro.AvroPucketDescriptor;
import com.intenthq.pucket.javacompat.PucketInstantiator;
import com.intenthq.pucket.util.ExceptionUtil;

import java.io.IOException;

public class AvroPucketInstantiator implements PucketInstantiator<IndexedRecord> {
  @Override
  public <T extends IndexedRecord> Pucket<T> newInstance(final Path path, final FileSystem fs, final String descriptor)
      throws RuntimeException, IOException {
    return ExceptionUtil.doThrow(
        AvroPucket.findOrCreate(path, fs, ExceptionUtil.doThrow(
            AvroPucketDescriptor.<T>apply(descriptor))
        )
    );
  }
}

File: avro/src/main/scala/com/intenthq/pucket/avro/mapreduce/AvroPucketInstantiator.scala
package com.intenthq.pucket.avro.mapreduce

import com.intenthq.pucket.Pucket
import com.intenthq.pucket.avro.{AvroPucket, AvroPucketDescriptor}
import com.intenthq.pucket.mapreduce.PucketInstantiator
import org.apache.avro.generic.IndexedRecord
import org.apache.hadoop.fs.{FileSystem, Path}

import scalaz.\/

/** Avro pucket instanciator for use with output format */
class AvroPucketInstantiator extends PucketInstantiator[IndexedRecord] {
  /** @inheritdoc */
  def newInstance[T <: IndexedRecord](path: Path, fs: FileSystem, descriptor: String,
                                      attempts: Int = Pucket.defaultCreationAttempts,
                                      retryIntervalMs: Int = Pucket.defaultRetryIntervalMs): Throwable \/ Pucket[T] =
    AvroPucketDescriptor[T](descriptor).flatMap(AvroPucket.findOrCreateRetry[T](path, fs, _, Pucket.defaultBlockSize, attempts, retryIntervalMs))
}
